### PR TITLE
Make MlflowClient serializable, cache utils._get_store

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Certain backend stores are known to be unserializable with pickle; for example, the SqlAlchemyStore cannot be pickled. However, the `MlflowClient` should still be pickleable regardless of backend store, as serializability of the `MlflowClient` is often required for distributed tracking workflows.

This PR makes `MlflowClient` serializable by ensuring that store objects are not set as properties / sub-properties on `MlflowClient` instances / any of their properties. This involves fetching stores via `[mlflow.tracking._tracking_service/mlflow.tracking._model_registry.utils]._get_store`; to ensure that fetching stores is performant (i.e. to ensure that we're not creating new instances every time a store is fetched), this PR introduces caching based on store URI (and artifact URI for Tracking).

## How is this patch tested?

Included unit tests

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/tracking/test_client.py [hunk 1]
```diff
@@ -500,3 +508,47 @@ def test_get_databricks_runtime_nondb(mock_spark_session):
...
+    mock_tracking_store = MockUnpickleableTrackingStore(
+        "sqlite:///" + backend_store_path, artifact_store_path
+    )
+    mock_model_registry_store = MockUnpickleableModelRegistryStore(
+        "sqlite:///" + backend_store_path
+    )
+
+    # Verify that the mock stores cannot be pickled because they are defined within a function
+    # (i.e. the test function)
+    with pytest.raises(AttributeError):
```
Comment (harupy) [inline tests/tracking/test_client.py:537]:
```suggestion
    with pytest.raises(AttributeError, match="<locals>.MockUnpickleableTrackingStore'"):
```

Can we specify `match` here to make sure pickle fails to dump the client due to `MockUnpickleableModelRegistryStore`?

File: tests/tracking/test_client.py [hunk 2]
```diff
@@ -500,3 +508,47 @@ def test_get_databricks_runtime_nondb(mock_spark_session):
...
+        pickle.dumps(mock_tracking_store)
+
+    with pytest.raises(AttributeError):
```
Comment (harupy) [inline tests/tracking/test_client.py:540]: same here

```suggestion
    with pytest.raises(AttributeError, match="<locals>.MockUnpickleableModelRegistryStore'"):
```

File: mlflow/tracking/_model_registry/client.py [hunk 3]
```diff
@@ -30,7 +30,10 @@ def __init__(self, registry_uri):
         :param registry_uri: Address of local or remote model registry server.
         """
         self.registry_uri = registry_uri
-        self.store = utils._get_store(self.registry_uri)
+
+    @property
+    def store(self):
+        return utils._get_store(self.registry_uri)
```
Comment (harupy) [inline mlflow/tracking/_model_registry/client.py:36]: nice trick!

File: tests/tracking/_tracking_service/test_utils.py [hunk 4]
```diff
@@ -196,6 +196,46 @@ def test_get_store_databricks_profile():
         assert "mycoolprofile" in str(e_info.value)
 
 
+def test_get_store_caches_on_store_uri_and_artifact_uri(tmpdir):
+    registry = mlflow.tracking._tracking_service.utils._tracking_store_registry
+
+    store_uri_1 = "sqlite:///backend_store_1.db"
+    store_uri_2 = os.path.join(str(tmpdir), "backend_store_2")
```
Comment (harupy) [inline tests/tracking/_tracking_service/test_utils.py:203]: Nit: Just feel free to ignore this comment :) I just wanted to share a simpler way (that I learned one month ago) to join file paths.

```suggestion
    store_uri_2 = tmpdir.join("backend_store_2").strpath
    # or
    # store_uri_2 = str(tmpdir.join("backend_store_2"))
```

File: mlflow/tracking/_model_registry/client.py (hunk 3 above)
Comment (harupy) [reply mlflow/tracking/_model_registry/client.py:36]: Now the store object is lazily initialized. It seems this behavior change broke some tests:

https://github.com/mlflow/mlflow/pull/4719/checks?check_run_id=3379318839#step:4:106

File: tests/tracking/_model_registry/test_utils.py [hunk 5]
```diff
@@ -102,6 +102,21 @@ def test_get_store_bad_uris(bad_uri):
         _get_store()
 
 
+def test_get_store_caches_on_store_uri():
+    store_uri_1 = "sqlite:///store1.db"
+    store_uri_2 = "sqlite:///store2.db"
```
Comment (harupy) [inline tests/tracking/_model_registry/test_utils.py:107]: Can we create these database files in a temporary directory?

File: tests/tracking/_tracking_service/test_utils.py [hunk 6]
```diff
@@ -196,6 +196,46 @@ def test_get_store_databricks_profile():
         assert "mycoolprofile" in str(e_info.value)
 
 
+def test_get_store_caches_on_store_uri_and_artifact_uri(tmpdir):
+    registry = mlflow.tracking._tracking_service.utils._tracking_store_registry
+
+    store_uri_1 = "sqlite:///backend_store_1.db"
```
Comment (harupy) [inline tests/tracking/_tracking_service/test_utils.py:202]: Same here (ref: https://github.com/mlflow/mlflow/pull/4719/files#r692768786)

File: tests/tracking/_tracking_service/test_utils.py [hunk 7]
```diff
@@ -196,6 +196,46 @@ def test_get_store_databricks_profile():
...
+
+    store11 = registry.get_store(store_uri_2, artifact_uri_2)
+    store12 = registry.get_store(store_uri_2, artifact_uri_2)
+    assert store11 is store12
+
+    assert store1 is not store3
+    assert store3 is not store5
+    assert store5 is not store7
+    assert store7 is not store9
+    assert store9 is not store11
```
Comment (harupy) [inline tests/tracking/_tracking_service/test_utils.py:236]: Can we simplify the code here using `itertools`?

```python
    stores_uris = ["sqlite:///backend_store_1.db", os.path.join(str(tmpdir), "backend_store_2")]
    artifact_uris = [
        None,
        os.path.join(str(tmpdir), "artifact_root_1"),
        os.path.join(str(tmpdir), "artifact_root_2"),
    ]

    stores = []
    for args in itertools.product(stores_uris, artifact_uris):
        store1 = registry.get_store(*args)
        store2 = registry.get_store(*args)
        assert store1 is store2
        stores.append(store1)

    assert all(x is not y for x, y in itertools.combinations(stores, 2))
```

File: mlflow/tracking/_model_registry/client.py (hunk 3 above)
Comment (dbczumar) [reply mlflow/tracking/_model_registry/client.py:40]: Great suggestion! I've implemented it to preserve existing behavior, which provides useful up-front validation outside of tests too.

File: tests/tracking/_model_registry/test_utils.py (hunk 5 above)
Comment (dbczumar) [reply tests/tracking/_model_registry/test_utils.py]: Done! Also applied this suggestion to other test cases added by this PR.

File: tests/tracking/_tracking_service/test_utils.py (hunk 6 above)
Comment (dbczumar) [reply tests/tracking/_tracking_service/test_utils.py]: Done!

File: tests/tracking/_tracking_service/test_utils.py (hunk 4 above)
Comment (dbczumar) [reply tests/tracking/_tracking_service/test_utils.py]: Great suggestion! I've adopted this throughout test cases added in this PR.

File: tests/tracking/_tracking_service/test_utils.py (hunk 7 above)
Comment (dbczumar) [reply tests/tracking/_tracking_service/test_utils.py]: Fantastic! Thank you! I've applied this suggestion.

File: tests/tracking/test_client.py (hunk 1 above)
Comment (dbczumar) [reply tests/tracking/test_client.py]: Done! Thanks!

File: tests/tracking/test_client.py (hunk 2 above)
Comment (dbczumar) [reply tests/tracking/test_client.py]: Done!

**Review by dbczumar — COMMENTED** [review]
@harupy Thank you for the review! I've addressed your comments!

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by harupy** [comment]
https://github.com/mlflow/mlflow/pull/4719/checks?check_run_id=3385574405#step:5:806

```python
@pytest.mark.large
def test_sparkml_estimator_model_log(tmpdir, spark_model_estimator):
    # Print the coefficients and intercept for multinomial logistic regression
    old_tracking_uri = mlflow.get_tracking_uri()
    cnt = 0
    # should_start_run tests whether or not calling log_model() automatically starts a run.
    for should_start_run in [False, True]:
        for dfs_tmp_dir in [None, os.path.join(str(tmpdir), "test")]:
            print("should_start_run =", should_start_run, "dfs_tmp_dir =", dfs_tmp_dir)
            try:
                
                🔥
                # `tracking_dir` doesn't change
                tracking_dir = os.path.abspath(str(tmpdir.join("mlruns")))
                mlflow.set_tracking_uri("file://%s" % tracking_dir)
                
　　　　　🔥
                # First iteration: `tracking_dir` is created during store initialization in the first epoch
                # Second iteration: cached store is used and `tracking_dir` is NOT created
                if should_start_run:
                    mlflow.start_run()
                artifact_path = "model%d" % cnt
                cnt += 1
                sparkm.log_model(
                    artifact_path=artifact_path,
                    spark_model=spark_model_estimator.model,
                    dfs_tmpdir=dfs_tmp_dir,
                )
                model_uri = "runs:/{run_id}/{artifact_path}".format(
                    run_id=mlflow.active_run().info.run_id, artifact_path=artifact_path
                )

                # test reloaded model
                reloaded_model = sparkm.load_model(model_uri=model_uri, dfs_tmpdir=dfs_tmp_dir)
                preds_df = reloaded_model.transform(spark_model_estimator.spark_df)
                preds = [x.prediction for x in preds_df.select("prediction").collect()]
                assert spark_model_estimator.predictions == preds
            finally:
                mlflow.end_run()
                mlflow.set_tracking_uri(old_tracking_uri)
                x = dfs_tmp_dir or sparkm.DFS_TMP
                shutil.rmtree(x)
                
                🔥
                # `tracking_dir` is removed here
                shutil.rmtree(tracking_dir)

```

**Pull request merged**